Add JSON-RPC support for VDSM

kahowell opened the pull request.

kahowell committed:
1353119: Add JSON-RPC support for VDSM

Had to fix xmlrpc on modern Python; M2Crypto appears to be broken in F27 at least...

kahowell commented:
FYI, this has some overlap with #125 and probably some conflicts with #124. I'll happily rebase against either.

kahowell commented:
I added a main function to one of the new modules for smoketest of VDSM over JSON-RPC that works like this:

    python -m virtwho.virt.vdsm.jsonrpc --cert /etc/pki/vdsm/certs/vdsmcert.pem --key /etc/pki/vdsm/keys/vdsmkey.pem $HOSTNAME 54321 Host.getVMList onlyUUID=False

jirihnidek commented on virtwho/virt/vdsm/stomp.py:
```diff
@@ -0,0 +1,165 @@
+                return  # success connecting
+            except socket.error as e:
+                log.warning('Unable to connect %s:%s: %s' % (self.host, self.port, unicode(e)))
+        raise ConnectionError('Unable to connect to %s:%s' % (self.host, self.port))
```
The `ConnectionError` is not defined nor imported.

kahowell replied:
Ah, it's a builtin in python3... whoops. I will switch to `IOError`.

jirihnidek commented on virtwho/virt/vdsm/stomp.py:
```diff
@@ -0,0 +1,165 @@
+        return frame
+
+    def subscribe(self, destination):
+        id = uuid.uuid4()
```
Variable `id` shadows built-in method.

jirihnidek commented on virtwho/virt/vdsm/stomp.py:
```diff
@@ -0,0 +1,165 @@
+        raise ConnectionError('Unable to connect to %s:%s' % (self.host, self.port))
+
+    def close(self):
+        for id in self.subscription_ids:
```
Variable `id` shadows built-in method.

jirihnidek commented on virtwho/virt/vdsm/vdsm.py:
```diff
@@ -119,29 +122,61 @@ def _secureConnect(self):
-                self.logger.exception(e)
-        # Try http version if ssl is off or fails
-        self.server = xmlrpclib.Server("http://localhost:%s" % self.management_port)
+            except Exception:
```
Broad exceptions are not good and the `Exception` has already been caught at line 139.

jirihnidek commented on virtwho/virt/vdsm/stomp.py:
```diff
@@ -0,0 +1,165 @@
+                        self.socket = self.ssl_context.wrap_socket(self.socket)
+                    else:
+                        self.socket = M2Crypto.SSL.Connection(self.ssl_context)
+                self.socket.connect(addr)
```
You use socket in blocking mode. When vdsm server is not running/available, then smoke test will be blocked forever and behavior of virt-who will be strange.

I would recommend to use module `requests` for communication with server. Module `requests` is already used in `rhevm.py`. Thus using of `requests` here would not create another rpm dependency. Using of bare `socket` has potential for many problems.

kahowell replied:
Unfortunately, STOMP is its own protocol, and does not use HTTP, so we can't use requests for this. I will investigate using a timeout.

jirihnidek commented on virtwho/virt/vdsm/stomp.py:
```diff
@@ -0,0 +1,165 @@
+import M2Crypto.SSL.Connection  # TODO make conditional for Python 3 compat
+
+log = logging.getLogger(__name__)
+
```
Please add here comment about `python-stomper` module. May be it will be possible to use it in future.

jirihnidek commented on virtwho/virt/vdsm/stomp.py:
```diff
@@ -0,0 +1,165 @@
+along with this program; if not, write to the Free Software
+Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+"""
+
```
Please add some doc string here with description of the module (what is STOMP, why is it used, ...).

jirihnidek commented on virtwho/virt/vdsm/stomp.py:
```diff
@@ -0,0 +1,165 @@
+Module for STOMP messaging, part of virt-who
+
+Copyright (C) 2018 Kevin Howell <khowell@redhat.com>
+
```
GPL should not be doc string but it should be comment.

jirihnidek commented on virtwho/virt/vdsm/jsonrpc.py:
```diff
@@ -0,0 +1,115 @@
+"""
+Module for JSON-RPC over STOMP, part of virt-who
+
+Copyright (C) 2018 Kevin Howell <khowell@redhat.com>
```
GPL license should not be part of doc string, but it should be in comment.

jirihnidek commented on virtwho/virt/vdsm/jsonrpc.py:
```diff
@@ -0,0 +1,115 @@
+import uuid
+
+from virtwho.virt.vdsm.stomp import StompClient
+
```
I suspect you aware of `python-jsonrpclib` module and I understand reasons why it is not used. Please add note (link) here about the module with explanation why it is not used.

jirihnidek reviewed (changes requested):
I can confirm that it just works :-). You are fast 👍 
On the other side I'm not happy that we cannot use existing modules e.g. `python-jsonrpclib`, but we probably cannot do anything about this. Please add some unit tests covering new code in `jsonrpc.py` and `stomp.py`. I also request some changes in code and I have some suggestions.

kahowell commented:
@jirihnidek I have addressed all feedback and rebased against master.

There are several ways to view the commits in the PR...

Both of these, you've seen the changes, but are good for sanity-checking the rebase:
 - `git diff 5fbb71b..568f131` shows the difference between the PR's original commits before rebase and after rebase
 - `git diff 87f9cfe..568f131` shows the changes the original commits rebased makes to master.

Following that, I made a few logical commits addressing difference pieces:
 - `git show 76c11c9` documentation changes
 - `git show 3b110c3` refactoring to address XML-RPC fallback and fixing XML-RPC to work on modern Python (F27)
 - `git show 11f0062` addition of timeout
 - `git show 9e8cf5d` addition of tests for JSON-RPC and STOMP code, plus a few fixes

jirihnidek commented on virtwho/virt/vdsm/stomp.py:
```diff
@@ -0,0 +1,242 @@
+                self.socket = socket.socket(family)
+                self.socket.settimeout(self.timeout)
+                if self.ssl_context:
+                    self.socket = self.ssl_context.wrap_socket(self.socket)
```
This raises error on RHEL 7, because `self.ssl_context.wrap_socket` is not available here :-(

kahowell replied:
Had logic in for this, and removed unintentionally. I added it back in latest commit.

jirihnidek commented on virtwho/virt/vdsm/vdsm.py:
```diff
@@ -107,46 +112,120 @@ def _get_local_vds_name(self, trusted_store_path):
+                    transport = SSL_Transport(ssl_context)
+                uri = 'https://%s:%s' % (addr, self.management_port)
+            else:
+                uri = 'http://%s:%s' % (addr, self.management_port)
```
It looks like that we are enabling unsecured connection to remote server here. On the other side it seems that `_xmlrpc()` is called only with `addr='localhost'`, when `ssl_context=None`.

jirihnidek commented on virtwho/virt/vdsm/vdsm.py:
```diff
@@ -107,46 +112,120 @@ def _get_local_vds_name(self, trusted_store_path):
-            try:
-                self.server = self._secure_connect()
+            ssl_context = self._create_ssl_context()
+            addr = self._get_local_vds_name(self.trust_store_path)
```
It seems that `addr` is get only from certificate and value of `server` option from configuration file is completely ignored. This logic was present in old xmlrpc code too. @cnsnyder, I don' think that this is expected behavior. Thoughts?

jirihnidek reviewed (changes requested):
I like your progress 👍 , but unfortunately it doesn't work on RHEL7.  Details are in comments.

jirihnidek reviewed (approved):
ACK, please squash and then it is ready for merge.

cnsnyder commented:
One request for the squash, please add the bug number (1353119) to the beginning of the commit message. Thank you!
